### ISSUE TITLE
Request: Facility for telling Spring Scheduling instrumentation not to trace individual @Scheduled methods

### ISSUE THREAD

**Issue by GrahamLea** [open-telemetry/opentelemetry-java-instrumentation#5008]
**Is your feature request related to a problem? Please describe.**
Spring Scheduling instrumentation can only be either on or off. A method `@Scheduled` to run every second gets traced every second, even if it does nothing interesting. There is no way (that I can tell) to ask the instrumentation not to trace individual methods.

**Describe the solution you'd like**
It'd be great to be able to annotate a `@Scheduled` method with something like `@NoTracing` and have it not be traced.

**Describe alternatives you've considered**
I have a hack where I use an aspect over `TaskScheduler.schedule*(..)` to unwrap the original `Runnable` from a `ScheduledMethodRunnable` if the wrapped runnable is a `ScheduledMethodRunnable` and I can see a specific annotation on the method.

**Comment by trask** [comment]
hi @GrahamLea! I recently added `code.*` span attribute to the quartz spans for a similar use case, so that you could use sampling to filter out specific noisy jobs (#4332).

It looks like we should do the same for spring scheduling spans.

Though that still requires writing an agent extension to add and configure such a sampler, see https://github.com/open-telemetry/opentelemetry-java-instrumentation/issues/1060#issuecomment-932369726 if you want to go down that route.

**Comment by GrahamLea** [comment]
I think it's interesting that in the few cases where I've seen similar filtering requests mentioned, the advice offered is often that a `Sampler` will be needed.

From my quick look at the code around `AgentInstaller`, the way it handles `otel.javaagent.exclude-classes` is to never instrument those classes to begin with.

I'm sure there's a number of factors to consider in deciding which way to implement these things, but I would have thought that, generally, if it's possible to not instrument something rather than to instrument it and then throw the data away, the first would probably be preferable. 🤷🏻‍♂️

**Comment by iNikem** [comment]
As said in [our documentation](https://github.com/open-telemetry/opentelemetry-java-instrumentation/blob/main/docs/suppressing-instrumentation.md#even-more-fine-grained-control):

> This option should not be used lightly, as it can leave some instrumentation partially applied,
which could have unknown side-effects.

**Comment by GrahamLea** [comment]
@iNikem Thanks. Would you mind helping me connect your highlighting of that para to what I've said above? Is it that the approach used by `exclude-classes`  - of not instrumenting code - is generally not a good idea (bc "unknown side-effects") and so `Sampler`s are the preferred way to solve these kinds of problems? Or something else?

**Comment by iNikem** [comment]
Yes, it is easier to break instrumentation by using `exclude-classes` than by using sampler. E.g. by excluding some classes from instrumentation you may break intra-process context propagation and so your traces are wrong. But if you can thoroughly test your application with `exclude-classes` and verify that resulting telemetry makes sense to you, then I don't see any reasons why you should not use it :)

**Comment by GrahamLea** [comment]
Hi @trask.
Old thread, but I'm just trying to put all the pieces together here to get this working in our app (and hopefully leave a papertrail for others).

Am I right in thinking that, to employ selective filtering of Spring `@Scheduled` tasks, I need to...
1. Use a version of the Agent that includes this change from Feb 2022:
https://github.com/open-telemetry/opentelemetry-java-instrumentation/pull/5306
2. And then should I...
    a. create my own extension with a custom sampler based off this example...:
https://github.com/open-telemetry/opentelemetry-java-instrumentation/blob/main/examples/extension/src/main/java/com/example/javaagent/DemoSampler.java
... and the `code.namespace` and `code.function` attributes to match the class and function names (based off):
https://github.com/open-telemetry/opentelemetry-java-instrumentation/blob/65717dae3e7d7ee2708e5e3a76744da82333ee2e/instrumentation-api-semconv/src/main/java/io/opentelemetry/instrumentation/api/instrumenter/code/CodeAttributesExtractor.java#L37-L43
    b. or should I just be building for myself the `RuleBasedRoutingSampler` from here?:
https://github.com/open-telemetry/opentelemetry-java-contrib/blob/main/samplers/src/main/java/io/opentelemetry/contrib/samplers/RuleBasedRoutingSampler.java
5. And then include that extension at runtime using `-Dotel.javaagent.extensions` as described here:
https://github.com/open-telemetry/opentelemetry-java-instrumentation/tree/main/examples/extension
6. At this point, I'm lost. How does the `Sampler` get instantiated and installed in the Agent? If it needs configuration (e.g. the rules for the `RuleBasedRoutingSampler`), how are they controlled by our app rather than needing to be baked into the extension JAR?

Any guidance you can give would be very appreciated! 🙏

**Comment by GrahamLea** [comment]
And on my configuration question, I see that the comments in `DemoSampler` have pointed me at [DemoAutoConfigurationCustomizerProvider](https://github.com/open-telemetry/opentelemetry-java-instrumentation/blob/main/examples/extension/src/main/java/com/example/javaagent/DemoAutoConfigurationCustomizerProvider.java)

**Comment by trask** [comment]
hi @GrahamLea! i'm not sure why the sampler is not getting called. I updated the example extension to use the (better) `addSamplerCustomizer` which you have used above, and it is working: #6204

**Comment by GrahamLea** [comment]
Thanks for the reply, Trask.
I'm using Honeycomb's customised OTel agent. I think the next step will have to be checking if it's overwriting my Sampler with one of it's own.

**Comment by GrahamLea** [comment]
Yep. Honeycomb's [HoneycombAutoConfigurationCustomizerProvider](https://github.com/honeycombio/honeycomb-opentelemetry-java/blob/main/custom/src/main/java/io/honeycomb/opentelemetry/HoneycombAutoConfigurationCustomizerProvider.java) calls `setSampler()`, and it's being loaded after my `AutoConfigurationCustomizerProvider`, so it blats my decorator-style sampler.
I can't see any way to prioritise AutoConfigurationCustomizerProviders so that mine is loaded 2nd. Do you know of any way to do that?
Seems like this is a problem with the way Honeycomb does its auto configuration. I'll raise with them.

**Comment by GrahamLea** [comment]
Okay, I've now got my Filter Sampler working (using a fork of the Honeycomb agent - hopefully temporary).

But... with a lot of these `@Scheduled` tasks that I want to ignore, the logic inside is...
```
once every few seconds {
    check some condition quickly
    if condition is true {  // occurs once every few minutes to hours
        do some significant piece of work
    }
}
```

When I get to `do some significant piece of work`, I **DO** want to trace what's going on there.
I've tried to do this by manually creating a `Span` in the code at that point, e.g. ...
```
        val span: Span = tracer.spanBuilder(spanName).startSpan()
        try {
            // do some significant piece of work
        } finally {
            span.end();
        }
```
But I'm not seeing those appear in my telemetry. (And my sampling is set to 100%)

Is that because it's within a `Span` that my filter is telling OTel to `DROP`?

Is there an easy way for me to start a new trace within a Span that's been generated by the agent instrumentation and then dropped by a `Sampler`?

**Comment by trask** [comment]
> Is that because it's within a `Span` that my filter is telling OTel to `DROP`?

yes, exactly. the default samplers are all "parent-based" which means they respect the decision of the parent (which includes the parent's decision to drop).

try changing your sampler to explicitly "record" the nested span instead of delegating to the underlying (parent-based) sampler

**Comment by GrahamLea** [comment]
Thanks. I need to check I'm understanding, and what the solution is...
The purpose of my filter Sampler is to `DROP` Spans that match a filter, because they happen too often.
But now I realise I want the option to have a Span created _manually_ within those dropped Spans that _will_ get sampled.

I believe what you're saying is I should change my filter to detect _those_ spans and give them a `RECORD` result instead? (Why not `RECORD_AND_SAMPLE`?)

So, is there an obvious way for me to detect when something is a manual span created under my DROPped spans? Or do I need to create one, e.g. add a custom Span attribute that tells my filter to record a Span instead of looking at the parent Span's sampling? Or should I just assume that if there's another INTERNAL under one of my DROPped spans that it's manual and I should sample it?

So many questions. Feel free to point me at some docs if this is explained well somewhere. (I tried searching, but all explanations I found of custom Samplers are pretty high level.)
